dds/DCPS/security/BuiltInPluginLoader.cpp

```diff
@@ -22,17 +23,19 @@ static const std::string PLUGIN_NAME("BuiltIn");
 int
 BuiltInPluginLoader::init(int /*argc*/, ACE_TCHAR* /*argv*/[])
 {
-  if (initialized) return 0;  // already initialized
-
-  SecurityPluginInst_rch plugin = DCPS::make_rch<BuiltInSecurityPluginInst>();
-  TheSecurityRegistry->register_plugin(PLUGIN_NAME, plugin);
-
-  SecurityConfig_rch default_config =
-    TheSecurityRegistry->create_config(SecurityRegistry::DEFAULT_CONFIG_NAME,
-                                       plugin);
-  TheSecurityRegistry->default_config(default_config);
-
-  initialized = true;
+  SecurityPluginInst_rch plugin = TheSecurityRegistry->get_plugin_inst(
+    PLUGIN_NAME, false /* Don't log error */);
+  if (!plugin) {
+    plugin = DCPS::make_rch<BuiltInSecurityPluginInst>();
+    TheSecurityRegistry->register_plugin(PLUGIN_NAME, plugin);
+  }
+
+  if (TheSecurityRegistry->has_no_configs()) {
+    SecurityConfig_rch default_config =
+      TheSecurityRegistry->create_config(SecurityRegistry::DEFAULT_CONFIG_NAME,
+                                         plugin);
+    TheSecurityRegistry->default_config(default_config);
+  }
```

Comment:
I guess if I'm creating a new `BuiltInPluginLoader` object, then there's not much of a point of doing all this.

Reply:
Please explain this a bit more

Reply:
So the issue is that the security config is getting removed from the security register, but for a static build it's not getting replaced when needed. In the restart test this happens after the first three Applications are finished but before the forth one was begun. During this point there are no Applications and I guess the security registry gets cleared.

At first I thought it was because the loader could only populate the security register once, but was still getting called multiple times. After I changed it to do it when when the security registry has no configs, it still wasn't working, because the function wasn't getting called again. It was only getting called when ACE was calling it at the beginning of the program. I didn't remove the changes I made to the loader though and just added a loader with a manual call to the restart test, which does work.

If the security plugin calls this automatically though, then this is needed and will eliminate the need to add a manual call in the user application, which is my question in the description of this PR. It might be easier to make that change and show that verses trying to explain it.

Reply:
> During this point there are no Applications and I guess the security registry gets cleared.

Small correction, I'm certain now the security registry singleton is getting destroyed.

> I guess if I'm creating a new BuiltInPluginLoader object, then there's not much of a point of doing all this.

I've had a change of heart on this. The change to `BuiltInPluginLoader` isn't doing much now, but assuming I've done it right I think I've made the function more robust. For example init was called again and the builtin plugin was still registered, but there were no configs set up for some reason, then this would go ahead and make sure a config existed.